### PR TITLE
592 intent proposal create interaction

### LINKED ISSUE #592: Intent Proposal: Create Interaction
### Use Case
An Analyst is IMing with a Buy Side fund manager who has a number of questions about the assumptions made in the Analyst’s latest report on a particular stock. There is a MiFIDII-style service agreement in place with the Buy Side institution and so the Analyst needs to record the fact that the questions were asked and a summary of the questions as a reportable interaction.

#### CreateInteraction
CreateInteraction should be used when a user wants to record a new Interaction in a system based on activity in another system as the starting point. CreateInteraction would be called with the Interaction context.

#### Interaction
An interaction is a significant direct exchange of ideas or information between a Sell Side party and one or more Buy Side parties. An interaction might be a call, a meeting (physical or virtual), an IM or the preparation of some specialist data.

#### Details
| Property    | Type    | Required | Example Value       |
|-------------|---------|----------|---------------------|
| `type`      | string  | Yes      | `fdc3.interaction` |
| `participants`      | fdc3.contactList  | Yes       | See below      |
| `timeRange` | fdc3.timeRange  | Yes       | See below            |
| `interactionType` | string  | Yes       | `Instant Message`            |
| `description`    | string  | Yes       | `Blah, blah, blah`         |
| `initiator`    | fdc3.contact  | No       | See below         |
| `origin`    |  string  | No       | `Outlook`         |

#### Example
```js
const interaction = {
    type: 'fdc3.interaction',
    participants: {
        type: 'fdc3.contactList',
        contacts: [
            {
                type: 'fdc3.contact',
                name: 'Jane Doe',
                id: {
                    email: 'jane.doe@mail.com'
                }
             },
            {
                type: 'fdc3.contact',
                name: 'John Doe',
                id: {
                    email: 'john.doe@mail.com'
                }
            },
        ]
    },
    interactionType: 'Instant Message',
    timeRange: {
        type: 'fdc3.timeRange',
        startTime: '2022-02-10T15:12:00Z'
    },
    description: 'Laboris libero dapibus fames elit adipisicing eu, fermentum, dignissimos laboriosam, erat, risus qui deserunt. Praesentium! Reiciendis. Hic harum nostrud, harum potenti amet? Mauris. Pretium aliquid animi, eget eiusmod integer proident. Architecto ipsum blandit ducimus, possimus illum sunt illum necessitatibus ab litora sed, nonummy integer minus corrupti ducimus iste senectus accumsan, fugiat nostrud? Pede vero dictumst excepturi, iure earum consequuntur voluptatum',
    initiator:  {
        type: 'fdc3.contact',
        name: 'Jane Doe',
        id: {
            email: 'jane.doe@mail.com'
        }
    },
    origin: 'Outlook'
}

fdc3.raiseIntent('CreateInteraction', interaction)
```

#### Additional Information
In this comment https://github.com/finos/FDC3/discussions/545#discussioncomment-2096489 I suggested the use of additional, optional Contexts. This would be a great example of where they could be used. The Interaction Context proposed gives everything required to log a significant interaction but additional information such as Instruments under discussion in the Interaction would only enrich the quality of data in the end system.

### PR DESCRIPTION
**Pull request by pauldyson**
Closes #592

### CODE REVIEW

**Comment by pauldyson** [comment]
@kriswest @mistryvinay the PR build is failing because we reference TimeRange (as discussed at the last working group) and whilst it is documented, the TS definition appears to be missing. I'm happy to add it but I'm assuming it should be in master rather than this branch.

**Comment by mistryvinay** [comment]
> @kriswest @mistryvinay the PR build is failing because we reference TimeRange (as discussed at the last working group) and whilst it is documented, the TS definition appears to be missing. I'm happy to add it but I'm assuming it should be in master rather than this branch.

Hi @pauldyson i think the issue you are seeing is because of this https://github.com/finos/FDC3/pull/724/files#r881622916 will let @kriswest comment if this is correct.

**Comment by Yannick-Symphony** [comment]
should this be called "record interaction"? reading through the PR code I got the impression that this was a generic way to start an interaction through various channels. but discussing with the domain experts it's more a case of recording an existing interaction to CRM or similar?

**Comment by kriswest** [comment]
@Yannick-Symphony I believe it's Create as in [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete).  It's following the naming conventions at https://fdc3.finos.org/docs/next/intents/spec#intent-name-prefixes (although looks like we went off-piste with 'Get' rather than 'Retrieve'.

Note that intents also have display names (although there is a slight problem with them as each app defines them #312 - it hasn't been high on the list to solve however so is still open).

**Comment by kriswest** [comment]
@pauldyson @milindchidrawar could you resolve conflicts - when you push it'll re-run the checks and should pass now.

File: docs/context/spec.md
Comment (kriswest) [inline docs/context/spec.md:151]: pls move after the fdc3.instrument* types
```suggestion
- [`fdc3.instrument`](ref/Instrument) ([schema](/schemas/next/instrument.schema.json))
- [`fdc3.instrumentList`](ref/InstrumentList) ([schema](/schemas/next/instrumentList.schema.json))
- [`fdc3.interaction`](ref/Interaction) ([schema](/schemas/next/interaction.schema.json))
```

File: src/context/schemas.json
Comment (kriswest) [inline src/context/schemas.json:14]: Alphabetical order
```suggestion
  "Instrument": ["https://fdc3.finos.org/schemas/next/instrument.schema.json"],
  "InstrumentList": ["https://fdc3.finos.org/schemas/next/instrumentList.schema.json"],
  "Interaction": ["https://fdc3.finos.org/schemas/next/interaction.schema.json"],
```

File: src/intents/Intents.ts
Comment (kriswest) [inline src/intents/Intents.ts:16]: Add in alphabetical order pls

File: src/intents/standard intents.json
Comment (kriswest) [inline src/intents/standard intents.json:61]: Add in alphabetical order pls

File: website/sidebars.json
Comment (kriswest) [inline website/sidebars.json:55]: Add in alphabetical order pls

File: website/sidebars.json
Comment (kriswest) [inline website/sidebars.json:75]: Alphabetical order pls
```suggestion
          "context/ref/Instrument",
          "context/ref/InstrumentList",
          "context/ref/Interaction",
```

File: website/yarn.lock
Comment (kriswest) [inline website/yarn.lock:7]: Minor, but probably best to remove file from PR

**Comment by kriswest** [comment]
Added suggestions for alphabetical order corrections, couldn't do them all unfortunately. 
If you do them here you can then merge this into your other PR #761  OR you can rebase that PR off master to remove this one's changes from it (technically better as it makes them independent and the diffs are nicer - however I think both will go in at around same time).

**Comment by mistryvinay** [comment]
hi @kriswest 
I also have this PR which is awaiting approval by @milindchidrawar @pauldyson to resolve some of the above. https://github.com/singletracksystems/FDC3/pull/7

**Comment by milindchidrawar** [comment]
Hi @kriswest @mistryvinay many thanks for the feedback. I will be working through these updates ahead of our next working group meeting and will let you know if I have any questions.

**Comment by kriswest** [comment]
@milindchidrawar don't forget to take a look at VInay's PR into this branch and resolve conflicts with master. It should then build ok.

**Comment by pierreneu** [comment]
Hi @pauldyson @milindchidrawar @kriswest Should we also add the ability to create multiple interactions at once?
E.g fdc3.raiseIntent('CreateInteraction', interaction**List**)

Just thinking about being able to report a conversation and not a single message from a chat application.

**Comment by pauldyson** [comment]
@pierreneu the intention is that it is a single interaction (conversation) whether it is one message or a whole series of them

**Comment by Yannick-Symphony** [comment]
when https://github.com/finos/FDC3/pull/779/ is merged, maybe we could extend "description" to take a string **or** fdc3.chat.message ?

File: CHANGELOG.md
Comment (kriswest) [inline CHANGELOG.md:49]: ViewInteractions already made it in in 2.0. Hence, this needs updating. Probably also worth merging master into the PR and resolving any conflicts @milindchidrawar @pauldyson

**Review by kriswest — CHANGES_REQUESTED** [review]
minor issue in changelog
Please merge master branch into PR and resolve any conflicts that arise.

File: CHANGELOG.md
Comment (milindchidrawar) [reply CHANGELOG.md]: Hi @kriswest - I've removed the reference to ViewInteractions. Hopefully that covers everything now.

File: CHANGELOG.md
Comment (kriswest) [inline CHANGELOG.md:59]: Apologies @milindchidrawar this should go into the unreleased section at the top of this file, as 2.0 was already released

File: docs/intents/ref/CreateInteraction.md
Comment (kriswest) [inline docs/intents/ref/CreateInteraction.md:21]: Missing the primary context for this intent:

```suggestion
* [ContactList](../../context/ref/ContactList)
* [Interaction](../../context/ref/Interaction)
```

File: src/context/ContextTypes.ts
Comment (kriswest) [reply src/context/ContextTypes.ts:123]: Also, this looks like a manual addition as there are other parts that are missing (martialling functions to/from JSON). 

That's not the end of the world as we'll re-generate this file after a merge to master. Hence, feel free to either correct this or remove the changes to this file entirely.

**Review by kriswest — CHANGES_REQUESTED** [review]
Theres still a small handful of minor changes to do here - but otherwise this looks good to me!

File: src/context/ContextTypes.ts
Comment (milindchidrawar) [reply src/context/ContextTypes.ts]: Hi @kriswest thanks for your patience with this. This should all be taken care of now.

File: website/static/schemas/next/interaction.schema.json
Comment (kriswest) [inline website/static/schemas/next/interaction.schema.json:36]:
```suggestion
  "required": ["participants", "timeRange", "interactionType", "description"]
```

**Comment by kriswest** [comment]
@milindchidrawar spotted another minor error in the schema - please commit the suggestions I've just added.

**Review by mistryvinay — COMMENTED** [review]
LGTM

**Comment by pierreneu** [comment]
> @pierreneu the intention is that it is a single interaction (conversation) whether it is one message or a whole series of them

I guess that's fine. So there is a single initiator to the interaction, and when we are reporting several messages as part of one interaction, the initiator field cannot be used to track the sender of each message. 
If we want to track who said what with the current format, we would need to add the user names directly within the Description field.

**Comment by kriswest** [comment]
@mistryvinay this PR has conflicts with master again, however, i would suggest just going ahead and starting off a consolidated PR (branch off master), merge this into it (may require @milindchidrawar to merge master and resolve conflicts again, apologies), then we can merge in others (e.g. TransactionResult, Symphony PRs) and submit that consolidated PR to SWG for inclusion.

**Pull request merged**
